src/PathOrderOptimizer.h

```diff
@@ -0,0 +1,591 @@
+        for(size_t i = 0; i < paths.size(); ++i)
+        {
+            const Path& path = paths[i];
+            if(path.is_closed)
+            {
+                for(const Point& point : *path.converted)
+                {
+                    line_bucket_grid.insert(point, i); //Store by index so that we can also mark them down in the `picked` vector.
+                }
+            }
+            else //For polylines, only insert the endpoints. Those are the only places we can start from so the only relevant vertices to be near to.
+            {
+                line_bucket_grid.insert(path.converted->front(), i);
+                line_bucket_grid.insert(path.converted->back(), i);
+            }
+        }
+
+        //For some Z seam types the start position can be pre-computed.
+        //This is faster since we don't need to re-compute the start position at each step then.
+        const bool precompute_start = seam_config.type == EZSeamType::RANDOM || seam_config.type == EZSeamType::USER_SPECIFIED || seam_config.type == EZSeamType::SHARPEST_CORNER;
+        if(precompute_start)
+        {
+            for(Path& path : paths)
+            {
+                if(!path.is_closed)
+                {
+                    continue; //Can't pre-compute the seam for open polylines since they're at the endpoint nearest to the current position.
+                }
+                if(path.converted->empty())
+                {
+                    continue;
+                }
+                path.start_vertex = findStartLocation(path, seam_config.pos);
+            }
+        }
+
+        std::vector<bool> picked(paths.size(), false); //Fixed size boolean flag for whether each path is already in the optimized vector.
+        Point current_position = start_point;
+        std::vector<Path> optimized_order; //To store our result in. At the end we'll std::swap.
+        optimized_order.reserve(paths.size());
+        while(optimized_order.size() < paths.size())
+        {
+            size_t best_candidate = 0;
+            coord_t best_distance2 = std::numeric_limits<coord_t>::max();
+
+            //First see if we already know about some nearby paths due to the line bucket grid.
+            std::vector<size_t> nearby_candidates = line_bucket_grid.getNearbyVals(current_position, 10);
+            std::vector<size_t> available_candidates;
+            available_candidates.reserve(nearby_candidates.size());
+            for(const size_t candidate : nearby_candidates)
+            {
+                if(picked[candidate])
+                {
+                    continue; //Not a valid candidate.
+                }
+                available_candidates.push_back(candidate);
+            }
+            if(available_candidates.empty()) //We may need to broaden our search through all candidates then.
```

Comment:
At some point we might want to consider a kd-tree that can do proper 'n nearest neighbour' searches.

Reply:
Yeah. Out of scope for now though, I'd say. It's not very slow yet.

Reply:
Fair enough  : - )